rdflib/plugins/sparql/evaluate.py

```diff
@@ -214,7 +214,10 @@ def evalPart(ctx, part):
             pass  # the given custome-function did not handle this part
 
     if part.name == 'BGP':
-        return evalBGP(ctx, part.triples)  # NOTE pass part.triples, not part!
+        # Reorder triples patterns based on 'bindedness' based on current ctx
+        triples = sorted(part.triples, key=lambda t: len([n for n in t if ctx[n] is None]))
```

Comment:
Your guess is spot on. I am not sure when this kicks in, for only "static" triples (with not optional/bindings/values/anything) - the code here: https://github.com/RDFLib/rdflib/blob/master/rdflib/plugins/sparql/algebra.py#L84 will already have sorted them correctly. 

This code helps when some other source of bindings has given us more bindings since then - in my base it was using initBindings.